### ISSUE TITLE
[Bug] select with round function leads to BE crash

### ISSUE THREAD

**Issue by panyanye** [apache/doris#33895]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

branch-2.0 (doris-2.0.7-dc065ead16)

### What's Wrong?

SELECT SQL with round funcrion leads to BE crash, because of NullPointerException.

The crash stack as follows:
![image](https://github.com/apache/doris/assets/57991859/ad8c9d5e-6208-478b-b437-859710e04440)

It looks like in file typeid_cast.h the function " To typeid_cast(From* from) " return null point.

### What You Expected?

be work normal not crash

### How to Reproduce?

create table test(big_key bigint)
DISTRIBUTED BY HASH(big_key) BUCKETS 1 PROPERTIES ("replication_num" = "1");

select 1 from test e1
having truncate(100,2)<-2308.57
and cast(round(round(465.56,min(-5.987)),2) as DECIMAL) in (
select cast(truncate(round(8990.65-4556.2354,2.4652),2)as DECIMAL)
from test r2);

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by Mryange** [comment]
The problem still exists on the master branch. Can we connect on WeChat for further discussion?
WeChatID  : ygg_0721_114514

**Comment by panyanye** [comment]
> The problem still exists on the master branch. Can we connect on WeChat for further discussion? WeChatID : ygg_0721_114514

OK，I have sent the application.

**Comment by zhiqiang-hhhh** [comment]
@Mryange how to reproduce the problem on master

**Comment by zhiqiang-hhhh** [comment]
@panyanye hello, is this sql generated by some test utils?

**Comment by panyanye** [comment]
> @panyanye hello, is this sql generated by some test utils?

No,it's a test SQL I wrote. The original SQL is somewhat complex, and this is a simplified version of it.

**Comment by zhiqiang-hhhh** [comment]
@panyanye the problem here is that we get a wrong plan from legacy planner (not the newest planner whose name is nereids), the incorrect plan changes decimal precision from Decimal64 to Decimal32, and makes execution engine get a null pointer. You can upgrade to 2.1.x, or just `set enable_fallback_to_original_planner = false` on branch 2.0, at least this can make your be not crash.

**Comment by zhiqiang-hhhh** [comment]
after we finished the tasks in https://github.com/apache/doris/issues/32755, the problem will be completely solved.

**Comment by panyanye** [comment]
> after we finished the tasks in #32755, the problem will be completely solved.

Ok, thank you very much~~~
